### PR TITLE
Python splines

### PR DESCRIPTION
**Pull request by paulstapor**
This thing is gonna to be spline support in Amici on the python side, therefore fixing #1198 .
However, before this becomes a valid PR, there are quite some things to be sorted out...
(everyone, feel free to add things to the list or discuss about the items, i.e., "I don't think that point xyz is important for merging" is also an appreciated contribution :) )
 - [ ] Adapt boundary condition and extrapolation in C++ code, according to python side
 - [ ] change handling in spline_constructors file (not returning a vector, but passing a vector of AbstractSpline pointers)
 - [x] Bugfix current extrapolation implementation on C++ side
 - [ ] Implement periodic boundary condition on C++ side
 - [ ] Add a "final value member" to spline class, to make it play nicely with steadystate solver (if spline is to be evaluated at `t = inf`, the "final value" member should simply be read.
 - [ ] Implement spline reinitialization, to make it work with preequilibration and presimulation
 - [x] Double-check example from lcontento which is still failing
 - [ ] Add docstrings (looks like hell at least on C++ side atm :imp: )
 - [ ] Allow splines also for species assignment rules (not only for parameters) and integrate with x_rdata in C++
 - [ ] Allow also formulae for spline slopes, i.e., don't force finite differences for Hermite splines
 - [ ] Implement tension parameter for Hermite splines? (i.e., allowing genera cardinal splines?)
 - [ ] Thinking carefully of which data members should be private, protected, public in splne class
 - [ ] Address sonarcloud code review 
 - [ ] Implementing C2-splines as other subclass?
 - [ ] Switching Matlab implementation to new spline class, 
 - [ ] Maybe reimplement JakStat-Swameye model with new spline class and add to benchmark collection?
 - [x] Implement natural boundary conditions in C++
 - [ ] Add a very extensive test suite, which checks:
     - [ ] Multiple splines with different length
     - [ ] Arbitrary formulae in spline nodes (may depend on p, k, and hard coded numeric values
     - [ ] Check splines  for species assignment rules
     - [ ] Formulae for spline slopes
     - [ ] Basically all this for all types of boundary conditions and extrapolations
     - [ ] All this with checking of `x`, `sx`, `y`, `sy`, `sllh`
     - [ ] Ideally all checked via parsing an Amici anotation, running through the full formula, and checking with analytically computed value

### CODE REVIEW

File: include/amici/symbolic_functions.h [hunk 1]
```diff
@@ -3,32 +3,242 @@
...
+ * negative values
+ *
+ * @param x argument
+ * @return if(x>0) then log(x) else -Inf
+ *
+ */
 double log(double x);
+
+/**
+ * C implementation of matlab function dirac
```
Comment (yannikschaelte) [inline include/amici/symbolic_functions.h:17]: I would not write "matlab" for these functions, as this is just an implementation of that function, not specific to the matlab side (same for heaviside, min, max, ...).

Comment (paulstapor) [reply include/amici/symbolic_functions.h:17]: Was just cut-copy-pasting old docstrings. Yep, would also be a good possibility to update them ;)

**Comment by dweindl** [comment]
Superseded by #1515
